Improved watching with chokidar

geigerzaehler opened the pull request:
This change implements the proposal in #3912 and improves the file watching behavior.

### Description of the Change

* We introduce the `--watch-files` command line option. This option allows control over which files are and is separate and more powerful than `--extension`. Fixes #2702.
* We introduce the `--watch-ignore` command line option that allows control over which files are not watched. Before this was hardcoded to `node_modules` and `.git`. See #2554.
* The `chokidar` package now handles file watching. (We’re using version `2.1.6` instead of `3.x` because the latter dropped support for Node v6.)
* New test files are picked up by the file watcher and run. Fixes #2176.

### Alternate Designs

Alternatives (like nodemon) were already discussed in #3912.

geigerzaehler commented:
I have two questions around this.

Should we deprecate the `--watch-extensions` command line option? How do I update the autogenerated `--help` output in `docs/index.md`?

juergba commented:
@geigerzaehler thank you for this PR!

I restarted some of the tests, they are still failing.
I will try to merge #3885 within the next few days, so we can use `chokidar 3.x`.

`--watch-extensions`: I don't know yet. Originally I was planning to keep it as an independent option (not just an alias of `--extension`). But if we have `--watch-files` with `glob`, we wouldn't need it anymore. Maybe for backwards compability.

--help output in docs/index.md: it's injected dynamically, have a look the deploy preview of netlify, it's all done.

geigerzaehler commented on docs/index.md:
```diff
@@ -1093,10 +1093,26 @@ Can be specified multiple times.
 
+### `--watch-files <file|directory|glob>`
+
+> _New in v6.3.0_
```
Yes, one might consider this a breaking change. I’d defer to you, the maintainers, to make that decision since I don’t know how mocha has handled these changes in the past.

outsideris replied:
@mochajs/core What do you think? 
It looks `semver-major` for me.

craigtaub replied:
So it's a backwards-incompatible change to `watch-extensions` (i.e. )? In which case major.

geigerzaehler replied:
> So it's a backwards-incompatible change to `--watch-extensions`?

The only thing that changed with respect to behavior of `--watch-extensions` is that newly created files are now watched, too. Otherwise the behavior is the same.

juergba replied:
What if `--watch-extensions` and `--watch-files` contradict each other?
I vote for removing `--watch-extensions` and `semver-major` v7.0.0.

geigerzaehler replied:
:+1:

juergba commented on docs/index.md:
```diff
@@ -1093,10 +1093,26 @@ Can be specified multiple times.
 Files having this extension will be considered test files. Defaults to `js`.
 
-Affects `--watch` behavior.
-
```
These extensions still affect `--watch` behavior, right? When `--watch-files` is `undefined`, files in the `cwd` with extension `opts.extension` will be watched.

geigerzaehler replied:
I didn’t want to confuse user by mentioning that it may affect `--watch` here. The behavior is well documented in `--watch-files`. I can restore it if you want.

juergba replied:
Ok. Then we should also remove watch from the Yargs option description, which currently is:
`--extension          File extension(s) to load and/or watch`

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/96c69f8ae17a5f9e6f57f69fc88bfd148de0e9f8

juergba commented on docs/index.md:
```diff
@@ -1093,10 +1093,26 @@ Can be specified multiple times.
+
+> _New in v6.3.0_
+
+List of comma separated paths or globs to watch when `--watch` is set. If a file matching the given If the argument is a glob a change to any file matching the glob will rerun the tests.
```
This sentence is weird.

juergba commented on docs/index.md:
```diff
@@ -1093,10 +1093,26 @@ Can be specified multiple times.
+
+If the path is a directory all files and subdirectories will be watched.
+
+By default all files in the current directory having on of the extensions provided by `--extensions` and not contained in the `node_modules` or `.git` folders are watched.
```
[...] having **one** of the extensions [...]
The option is `--extension` instead of `--extensions`.

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/ba7667e29c72589ec50ed00411fe272b3071808b

juergba commented on example/config/.mocharc.js:
```diff
@@ -12,5 +12,7 @@ module.exports = {
   timeout: 2000,
-  ui: 'bdd'
+  ui: 'bdd',
+  watchFiles: ['lib/**/*.js', 'test/**/*.js'],
```
Probably it's better to use the canonical names as shown in the documentation:
`watch-files`, `watch-ignore`

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/f18a0b66b5df455fc7b1598797e0e3245df52826

juergba commented on example/config/.mocharc.json:
```diff
@@ -10,5 +10,7 @@
   "timeout": 2000,
-  "ui": "bdd"
+  "ui": "bdd",
+  "watchFiles": ["lib/**/*.js", "test/**/*.js"],
```
same

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/f18a0b66b5df455fc7b1598797e0e3245df52826

juergba commented on example/config/.mocharc.jsonc:
```diff
@@ -10,5 +10,7 @@
   "timeout": 2000,
-  "ui": "bdd"
+  "ui": "bdd",
+  "watchFiles": ["lib/**/*.js", "test/**/*.js"],
```
Here you could leave it as is, and add a comment, that camelcased versions do work as well.

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/f18a0b66b5df455fc7b1598797e0e3245df52826

juergba commented on example/config/.mocharc.yml:
```diff
@@ -44,3 +44,8 @@ trace-warnings: true # node flags ok
 ui: bdd
 v8-stack-trace-limit: 100 # V8 flags are prepended with "v8-"
 watch: false
+watchFiles:
```
same

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/f18a0b66b5df455fc7b1598797e0e3245df52826

juergba commented on lib/cli/run.js:
```diff
@@ -241,6 +241,20 @@ exports.builder = yargs =>
         group: GROUPS.FILES
+      },
+      'watch-files': {
+        description: 'Comma separated list of paths or globs to watch',
```
I'm unsure about "Comma separated list". Normally (see eg `--file`) we prefer to repeat the option.
I will do some testing ...

geigerzaehler replied:
I’m fine with either

juergba replied:
Please change to "List of paths or globs to watch"

juergba commented on package.json:
```diff
@@ -514,6 +514,7 @@
   "dependencies": {
     "ansi-colors": "3.2.3",
     "browser-stdout": "1.3.1",
+    "chokidar": "2.1.6",
```
In v7.0.0 we drop Node v6, so `chokidar` v3 is ok.

geigerzaehler replied:
Updated in https://github.com/mochajs/mocha/pull/3980/commits/f18a0b66b5df455fc7b1598797e0e3245df52826

geigerzaehler commented on lib/cli/run-helpers.js:
```diff
@@ -124,7 +124,9 @@ exports.runMocha = (mocha, options) => {
-    spec = []
+    spec = [],
+    watchFiles,
+    watchIgnore
```
> No defaults here?

We don’t need defaults here. The default for `watchIgnore` is defined by `yargs` in `lib/cli/run`. For `watchFiles` we need `undefined` to signal the fallback to `--extensions`.

juergba commented on lib/cli/watch-run.js:
```diff
@@ -17,14 +17,42 @@ const collectFiles = require('./collect-files');
-module.exports = (mocha, {ui}, fileCollectParams) => {
-  let runner;
-  const files = collectFiles(fileCollectParams);
+module.exports = (mocha, {ui, watchFiles, watchIgnore}, fileCollectParams) => {
```
My IDE warns: 'ui' is declared but its value is never read.
All three properties of `obj` are already enclosed in the `mocha`instance. Do we need to pass this `obj` at all?

geigerzaehler replied:
Indeed, `ui` is unused. I removed it in https://github.com/mochajs/mocha/pull/3980/commits/10f4fda066ca96a242a3a69d52207e70664f1a01.

For the other options I’d favor passing them explicitly. They are only available through `mocha.options` because we dump all options there but it’s not clear whether this is an internal property or not. I tried isolate all the access to internal mocha properties in `resetMocha` (which should probably be turned into a method at some point).

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,30 @@ Files specified using `--file` _are not affected_ by this option.
 
+### `--watch-files <file|directory|glob>`
+
+> _New in v7.0.0
```
The final "_" is missing: `_New in v7.0.0_`

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/10f4fda066ca96a242a3a69d52207e70664f1a01

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,30 @@ Files specified using `--file` _are not affected_ by this option.
+
+### `--watch-ignore <file|directory|glob>`
+
+> _New in v7.0.0
```
dito

geigerzaehler replied:
Fixed in https://github.com/mochajs/mocha/pull/3980/commits/10f4fda066ca96a242a3a69d52207e70664f1a01

juergba commented:
Regarding time interval: I remember an issue which I couldn't find anymore. The user complained about his cpu being close to a collapse. He let Mocha run over night in watch mode with frequent file changes via web.
Do we need such an interval value?

geigerzaehler commented:
> Regarding time interval: I remember an issue which I couldn't find anymore. The user complained about his cpu being close to a collapse. He let Mocha run over night in watch mode with frequent file changes via web.
> Do we need such an interval value?

I’m not sure what you are referring to. Could you give some context? Do you mean the interval that we are polling the file system at? This can be [configured through chokidar](https://github.com/paulmillr/chokidar#performance).

juergba commented on lib/cli/run-option-metadata.js:
```diff
@@ -18,9 +18,11 @@ exports.types = {
+    'require',
+    'spec',
+    'watch-files',
+    'watch-ignore'
```
Something is wrong with `loadOptions()` in "cli/options". I have to check the camel-case option of yargs-parser.
When I use `--watch-files <path>` the result is - as it should be - an `array` , when I use `--watchFiles <path>` it's a `string`.

juergba replied:
Test case:
- input via CLI: `--watchFiles ./path1 --watchFiles ./path2`
- plus input via `.mocharc.yml`: `watchFiles: ['./path3']`

current result: `{watchFiles: [ "./path1", "./path2"]`
Makes sense, since the type of `watchFiles` is `undefined`, not `array`.

Setting `camel-case-expansion` to `true` (in "cli/options"), it's still wrong:
result: `{watch-files: [ "./path1", "./path2"], {watchFiles: [ "./path1", "./path2"]}`
This output seems to be a yargs-parser bug, since an `array`typed option should combine the result.

When I use `--watch-files` instead of `--watchFiles` the output is correct:
result: `{watch-files: [ "./path1", "./path2, "./path3"]}`

So the result is due to our camel-case settings and also due to a yargs-parser bug. I don't want to set a quick shot. Do you have an idea, otherwise let's go on and take care of this later.

geigerzaehler replied:
> Do you have an idea

Unfortunately I don’t. I’m not familiar with  yargs’ configuration loading.

juergba replied:
I will open a separate issue for this parsing issue.

juergba commented on lib/cli/watch-run.js:
```diff
@@ -54,29 +94,59 @@ module.exports = (mocha, {ui}, fileCollectParams) => {
+  mocha.unloadFiles();
+  mocha.suite = mocha.suite.clone();
+  mocha.suite.ctx = new Context();
+  mocha.ui(mocha.options.ui);
```
I don't see the meaning of this. We had a similar situation with `grep` in another PR.
`ui` is set by the `Mocha` constructor and is read once when Mocha is launched. There is no need to reset it here again.

geigerzaehler replied:
I added a comment in https://github.com/mochajs/mocha/pull/3980/commits/4e7c7437cc1c3348ab8e6b2f562f6ea1924e5f9e that explains why we need it. If you look at the code of `mocha.ui()` you see that it registers some hook on `mocha.suite` that is makes the global DSL (e.g. `describe`) work. If we omit the call to `mocha.ui()` the new test suites will never be registered.

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,30 @@ Files specified using `--file` _are not affected_ by this option.
-> _Updated in v6.0.0. Previously `--watch-extensions`, but now expanded to affect general test file loading behavior. `--watch-extensions` is now an alias_
+### `--extension <ext>`
 
 Files having this extension will be considered test files. Defaults to `js`.
```
> --global can accept a comma-delimited list; --global app,YUI is equivalent to --global app --global YUI.

Please add the equivalent information for `--extension`.

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/b422fa8f8ecc2d01beecc62e72f7463c2dba8075

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,30 @@ Files specified using `--file` _are not affected_ by this option.
+
+> _New in v7.0.0_
+
+List of comma separated paths or globs to watch when `--watch` is set. If a file matching the given glob changes or is added or removed mocha will rerun all tests.
```
Our main format for `array` typed options is repeating the flag. In some cases we also allow comma separated lists of values.
Therefore I prefer: "List of paths or globs to watch [...]"
then please add: "--watch-files can accept a comma-delimited list; --watch-files ... is equivalent to --watch-files ..."

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/b422fa8f8ecc2d01beecc62e72f7463c2dba8075

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,30 @@ Files specified using `--file` _are not affected_ by this option.
+
+> _New in v7.0.0_
+
+List of comma separated paths or globs to exclude from watching. Defaults to `node_modules` and `.git`.
```
dito

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/b422fa8f8ecc2d01beecc62e72f7463c2dba8075

juergba replied:
This one got lost: "List of paths or globs to exclude [...]"

geigerzaehler replied:
Done

juergba commented on lib/cli/run.js:
```diff
@@ -241,6 +241,20 @@ exports.builder = yargs =>
+      },
+      'watch-ignore': {
+        description:
+          'Comma separated list of paths or globs exclude from watching',
```
dito

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/b422fa8f8ecc2d01beecc62e72f7463c2dba8075

juergba commented on lib/cli/run-helpers.js:
```diff
@@ -137,7 +139,7 @@ exports.runMocha = (mocha, options) => {
 
   if (watch) {
-    watchRun(mocha, {ui}, fileCollectParams);
+    watchRun(mocha, {ui, watchFiles, watchIgnore}, fileCollectParams);
```
`ui` is not required here.

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/1fd64f8821a17ccb5296d6d31f6125a100e7b0bd

juergba commented on lib/cli/watch-run.js:
```diff
@@ -54,29 +94,59 @@ module.exports = (mocha, {ui}, fileCollectParams) => {
+ * @private
+ */
+const resetMocha = mocha => {
+  mocha.unloadFiles();
```
The `require.cache` is deleted twice, once in `beforeRun()` and the second time here.

juergba replied:
Ok, I got it. In `beforeRun()` the watched files are cleared, in `mocha.unloadFiles()` the test files are removed out of cache.

juergba commented on lib/cli/watch-run.js:
```diff
@@ -54,29 +94,59 @@ module.exports = (mocha, {ui}, fileCollectParams) => {
+};
+
+/**
+ * Return the list of absolute paths watchched by a chokidar watcher.
```
just typo "watchched"

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/2ce5e40b5dd74b75947ea23e843af390c21c13d7

juergba commented on lib/cli/watch-run.js:
```diff
@@ -35,17 +62,30 @@ module.exports = (mocha, {ui}, fileCollectParams) => {
-      mocha.files = files;
-      runAgain = false;
+      beforeRun();
+      resetMocha(mocha);
```
Even for the first run - on the `ready` event - we reset everything, right?

geigerzaehler replied:
Yes. This is not ideal but it avoids special casing the initial run.

juergba commented on lib/cli/watch-run.js:
```diff
@@ -54,29 +94,59 @@ module.exports = (mocha, {ui}, fileCollectParams) => {
     } else {
       rerun();
     }
+  };
```
If one of the watched files is changed while the tests are still running, the tests are aborted, but not rerun?
Is this the usual behavior of a watch functionality to abort, instead of waiting for the test end and then restarting?

Out of topic: I don't trust `runner.abort()`. I doubt that the hook pattern is correct and all `afterEach` / `after` hooks run for correct tear-down. This is prone to errors.

geigerzaehler replied:
This is the same behavior as before. I think it makes sense to abort the tests when a file changes. If you usually want to get quick feedback when your code changes.

juergba commented on lib/utils.js:
```diff
@@ -53,27 +53,6 @@ exports.isString = function(obj) {
-    });
-  });
-};
-
```
`utils.files` is not referenced anymore, so can be deleted as well.

geigerzaehler replied:
Awesome! Fixed in https://github.com/mochajs/mocha/pull/3980/commits/55df5546a53a5f9cbf23db54423ef3a6bfec5e04

juergba commented on test/integration/file-utils.spec.js:
```diff
@@ -1,152 +0,0 @@
-    }
-  });
-
-  describe('.lookupFiles', function() {
```
This function still exists, please don't delete the corresponding tests.

geigerzaehler replied:
Oops. I restored the test cases

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,36 @@ Files specified using `--file` _are not affected_ by this option.
+
+By default all files in the current directory having one of the extensions provided by `--extension` and not contained in the `node_modules` or `.git` folders are watched.
+
+The option can be given multiple times. The option accepts a comma-delimited list: `--watch-filas a,b` is equivalent to `--watch-files a --watch-files b`
```
typo: `--watch-files a,b`

geigerzaehler replied:
Done

geigerzaehler commented:
I think I fixed everything, rebased, and squashed the commits!

juergba commented on test/integration/options/watch.spec.js:
```diff
@@ -49,6 +138,25 @@ describe('--watch', function() {
       });
     });
 
+    it('picks up new test files when they are added', function() {
```
This test is identical to the one above.

geigerzaehler replied:
Done

juergba commented on docs/index.md:
```diff
@@ -1087,16 +1087,36 @@ Files specified using `--file` _are not affected_ by this option.
 
+The option can be given multiple times. The option accepts a comma-delimited list: `--extension a,b` is equivalent to `--extension a --extension b`
+
+### `--watch-files <file|directory|glob>`
```
Please move `--watch-files <file|directory|glob>` and `--watch-ignore <file|directory|glob>` documentation below `--watch, -w`. Just the same order as in the help output.

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/7cccd07725a9a72b1708b4983b30226cf3997c96

juergba replied:
There is one empty line too much before "--fgrep" (L1159), which results in a linting error.

juergba commented on lib/cli/run.js:
```diff
@@ -241,6 +241,19 @@ exports.builder = yargs =>
+        group: GROUPS.FILES,
+        requiresArg: true,
+        coerce: list,
+        default: ['node_modules', '.git']
```
Please put this default value in "lib/mocharc.json" and set `default: defaults.watchIgnore`.
Same procedure as `opts` option.

geigerzaehler replied:
https://github.com/mochajs/mocha/pull/3980/commits/f938922b10118ea7f9d58febffb4dec2d54bd0e6

juergba commented:
While testing I found this bug:
`node bin\mocha --watch ../skipped.js`

This should restart my test "skipped.js" when a file in `cwd` with extension "js" is touched.
This works with "karma.conf.js" and others.
It does not work with ".eleventy.js" or other files including a "." in its name.

juergba commented:
`node bin\mocha --watch notExistFile.js`

When I produce an error, eg. "Error: No test files found: notExistFile.js", the cursor remains hidden after the error message.

So that's it, no more testing from my side.

geigerzaehler commented:
> `node bin\mocha --watch notExistFile.js`
> 
> When I produce an error, eg. "Error: No test files found: notExistFile.js", the cursor remains hidden after the error message.

Fixed.

> While testing I found this bug:
> `node bin\mocha --watch ../skipped.js`
> 
> This should restart my test "skipped.js" when a file in `cwd` with extension "js" is touched.
> This works with "karma.conf.js" and others.
> It does not work with ".eleventy.js" or other files with a leading dot "." in its name.

Fixed in 0e2ee09e46c7019d3510f53a30f448323dd731d1

juergba commented:
@geigerzaehler thanks for your latest fixes! 👍

@boneskull @craigtaub @outsideris 
I'm going to merge this PR within the next few days.

jedwards1211 commented:
I just experienced what @juergba reports.  I'm using `@babel/register` and the `--watch-files` approach probably just doesn't work with require hooks.

geigerzaehler commented:
I was unable to reproduce the bug that @juergba mentioned. If this is still a problem, could somebody open an issue?

@jedwards1211: I’m happy to take a look at this but I need more information to investigate. For example the commands you are running, the expected behavior and the actual behavior. Feel free to open an issue and tag me.

juergba commented:
I think we have already an open issue for this: #4016 .
But this improved watch mode hasn't been introduced with the latest version of Mocha, it's more than an year ago.

jedwards1211 commented:
I'll see if I can make a repo to reproduce it tomorrow. I know this change isn't recent, but I just now noticed that `--watch-extensions` no longer works and tried using `--watch-files` in my project.

jedwards1211 commented:
Sigh, it was happening for me the other day, now I can't get it to reproduce, I hate when this happens.  I  was seeing behavior exactly like you described, "changes made on watched source files only trigger a new test run without considering the changes."  If it happens again I'll try to figure out why

jimmywarting commented:
question: is chokidar still better than builtin `fs.watch`?
